### ISSUE TITLE
3DS adds a large amount of whitespace to every ScrollView with SwiftUI and UIViewControllerRepresentable

### ISSUE THREAD

**Issue by AlexanderTheLit** [stripe/stripe-ios#1761]
## Summary
Setup Intent in a ViewController called by a UIViewControllerRepresentable in SwiftUI works great and integrates well. The only issue is that when I use 3DS the WebView causes this weird issue where large amounts of white space are added to every ScrollView in my app where buttons can't be tapped and other strangeness occurs like lists being clipped into a small rectangle. 

I have included the SwiftUI portion of the code below to reproduce, as for the Setup Intent ViewController I use this code exactly [https://github.com/stripe-samples/mobile-saving-card-without-payment/blob/master/client/ios/App%20(Swift)/CheckoutViewController.swift](url)

*What's weird is that it only occurs on device. I've tried with iPhone 12, XR, and 8 plus

## Code to reproduce

Use the card number 4000002500003155. Once the 3DS presents, tap fail or complete authentication. Then close the setup intent sheet and scroll down to see that the view has a large amount of whitespace below the blue Rectangle.

```
import SwiftUI

struct SetupIntent: UIViewControllerRepresentable {
    
    func makeUIViewController(context: Context) -> UIViewController {
        let input = SetupIntentViewController()
        return input
    }

    func updateUIViewController(_ uiViewController: UIViewController, context: Context) {
        
    }
    
}

struct ContentView: View {
    
    @State var isPresentingSetupIntent = true
    
    var body: some View {
        ScrollView{
            VStack{
                
                Button("Setup Intent"){
                    isPresentingSetupIntent = true
                }
                
                Rectangle()
                    .fill(Color.blue)
                    .frame(width: 400, height: 1000)
        
            }
            
        }.sheet(isPresented: $isPresentingSetupIntent, content: {
            SetupIntent()  
        })
    }
}
```

## iOS version
IOS 14.4
## Installation method
Pod

## SDK version
  - Stripe (21.2.0):
    - Stripe/Stripe3DS2 (= 21.2.0)
  - Stripe/Stripe3DS2 (21.2.0)

**Comment by csabol-stripe** [comment]
Hi @AlexanderTheLit, I tried to reproduce this issue on an iPhone X running both 14.2 and 14.4 but have not been able to. Do you have additional screenshots, videos, and/or example projects that you could share?

**Comment by AlexanderTheLit** [comment]
Thanks for the quick reply @csabol-stripe. Huh strange, I've included a video highlighting the issue on a iPhone XR 14.4. It only happens with the 3DS WebView, for a non 3DS SetupIntent it works perfectly. I have also included the SetupIntentViewController swift file I'm using.

[StripeViewController.swift.zip](https://github.com/stripe/stripe-ios/files/5955084/StripeViewController.swift.zip)

https://user-images.githubusercontent.com/78665850/107445841-3ef52e80-6b0b-11eb-9fb5-f440082c9fec.mp4

**Comment by csabol-stripe** [comment]
ah I am able to reproduce with your code. This also reproduces if you change the pay button to just present an SFSafariViewController instance (bypassing the Stripe SDK)
`@objc func pay() {
            let vc = SFSafariViewController(url: URL(string: "https://www.stripe.com")!)
            self.present(vc, animated: true, completion: nil)
    }`

It looks like this is probably an issue with SwiftUI + view controller presentation though I haven't been able to track down any similar bugs. I will leave this issue open for now because @davidme-stripe is working on some SwiftUI support and sample code now that may be able to help with this

**Comment by davidme-stripe** [comment]
@AlexanderTheLit, can you try the new SwiftUI bindings in 21.3.0? We now have an `STPPaymentCardTextField.Representable` SwiftUI View and a `.setupIntentConfirmationSheet()` ViewModifier, which should handle this use case. You can see a usage example in the [IntegrationTester](https://github.com/stripe-ios/stripe-ios/blob/private/IntegrationTester/IntegrationTester/Examples/CustomCardSetupIntent.swift) app.

**Comment by AlexanderTheLit** [comment]
Thanks for the reply @davidme-stripe. I integrated your code, but sadly the issue is still persisting. I think @csabol-stripe is correct, the issue lies with SwiftUI and SFSafariViewController presentation.

**Comment by davidme-stripe** [comment]
That does look like a SwiftUI bug, unfortunately. :( Please file it via [https://developer.apple.com/bug-reporting/](Feedback Assistant), we'll file our own report as well.

As for the sheet-in-a-sheet dismissing issue, that's a bug on our end. I'll investigate a fix.

**Comment by mvarie** [comment]
@AlexanderTheLit 
When you change the `modalPresentationStyle` of `SFSafariViewController` to `.overCurrentContext` or `.overFullScreen`, the problem disappears:

```
@objc func pay() {
  let vc = SFSafariViewController(url: URL(string: "https://www.stripe.com")!)
  vc.modalPresentationStyle = .overCurrentContext
  self.present(vc, animated: true, completion: nil)
}
```    

However, when using Stripe, you're not handling SFSafariViewController directly. 
Not that I recommend it, but to explore the issue further, the following extension seems to work when using Stripe (that is, it seems to make your problem disappear).

```
extension SFSafariViewController{
    
    open override var modalPresentationStyle: UIModalPresentationStyle{
        get{
            return .overCurrentContext
        }
        set {}
    }    
}

```

Of course, it's hacky at best.

**Comment by AlexanderTheLit** [comment]
Very interesting @mvarie, I just tried it and it does indeed fix the issue. To be honest I kind of like the way `.overCurrentContext` looks as-well. I'm going to mess around with it more tonight, but I'm probably going to push an update out with this because I've already had a few customers mention this issue.

**Comment by davidme-stripe** [comment]
This should be fixed in [21.4.0](https://github.com/stripe/stripe-ios/releases/tag/21.4.0).
